Dev ops

AlexanderNaide opened the pull request:
В данном пуле распиливал представления на вложенные представления (view внутри view). Мне показалось - это мне нужно))). Еще сюда попала работа с меню сайта: отработал механизм перестройки ссылок header в зависимости от того, в какой части сайта пользователь находится в данный момент.

AlexanderNaide committed:
пробуем state

AlexanderNaide committed:
*

AlexanderNaide committed:
***

AlexanderNaide committed:
***

AlexanderNaide committed:
***

AlexanderNaide committed:
***

AlexanderNaide committed:
***

AlexanderNaide committed:
меню закончено

AlexanderNaide committed:
меню закончено

AlexanderNaide committed:
Разобрался с ui-router.

AlexanderNaide committed:
Merge pull request #4 from AlexanderNaide/multiModulFronState

Multi modul fron state

AlexanderNaide committed:
Merge pull request #5 from AlexanderNaide/userAccaunt

User accaunt

AlexanderNaide committed:
Начинаем Кабинет пользователя. Расписание.

AlexanderNaide committed:
мелкие правки во время лекции

AlexanderNaide committed:
запрос на расписание

AlexanderNaide committed:
Бэк для расписания готов

AlexanderNaide committed:
Расписание. Фронт. Продолжение

AlexanderNaide committed:
запрос на расписание

AlexanderNaide committed:
запрос на расписание

AlexanderNaide committed:
запрос на расписание

AlexanderNaide committed:
добился таки корректного запуска изотопа

AlexanderNaide committed:
Добавлен период

AlexanderNaide committed:
***

AlexanderNaide committed:
***

AlexanderNaide committed:
Merge pull request #7 from AlexanderNaide/userAccaunt

User accaunt

AlexanderNaide committed:
начинаем кабинет админа

AlexanderNaide committed:
***

comshepherd commented on BackService/src/main/java/org/satal/backservice/controllers/UserController.java:
```diff
@@ -20,5 +20,12 @@ public class UserController {
+    @GetMapping("/schedule")
+    public SpecializationDto getSchedule(@RequestParam Long id){
+        Optional<Specialization> specializationOptional = specializationService.findById(id);
+        return specializationOptional.map(SpecializationDto::new).orElseGet(() -> new SpecializationDto(null, "Специализация с id:" + id + " не найдена."));
```
Необязательно:
1. лучше кидать исключение
2. ловить исключение в @ControllerAdvice и возвращать ответ с 404 кодом

AlexanderNaide replied:
Ой, это что-то временное было прикручено, обычно я так не делаю, исправляю.

comshepherd commented on BackService/src/main/java/org/satal/backservice/controllers/WorkoutController.java:
```diff
@@ -0,0 +1,44 @@
+//    }
+
+    @GetMapping("/week")
+    public WorkoutFullDto upAllUsers(@RequestParam(required = false, defaultValue = "0") Integer delta, Principal principal){
```
название метода `upAllUsers ` скорее всего осталось от предыдущего

comshepherd commented on BackService/src/main/java/org/satal/backservice/dto/worcouts/WorkoutFullDto.java:
```diff
@@ -0,0 +1,59 @@
+    private Map<String, String> classType;
+    private List<WorkoutWeekDto> week;
+
+    public WorkoutFullDto(Integer delta, List<String> list) {
```
`list ` как название переменной неинформативно, лучше давать более осмысленные имена

comshepherd commented on BackService/src/main/java/org/satal/backservice/dto/worcouts/WorkoutFullDto.java:
```diff
@@ -0,0 +1,59 @@
+    private String period;
+    private Integer currentWeek;
+//    private List<String> classType;
+    private Map<String, String> classType;
```
будет удобнее, если key типизировать enum'ом

comshepherd commented on BackService/src/main/java/org/satal/backservice/dto/worcouts/WorkoutFullDto.java:
```diff
@@ -0,0 +1,59 @@
+            String[] split = s.split(",");
+            classType.put(split[2], "." + split[2]);
+            WorkoutWeekDto w = week.stream().filter(e -> e.getDayOfWeek().equals(split[1])).findFirst().orElse(null);
+            assert w != null;
```
лучше кидать исключение с осмысленным текстом

comshepherd commented on BackService/src/main/java/org/satal/backservice/services/ImplementService.java:
```diff
@@ -24,6 +27,8 @@ public class ImplementService {
 
+    private final WorkoutService workoutService;
+
     public void init(){
```
Если эти данные не для разработки или тестов, то это может быть в миграциях БД

AlexanderNaide committed:
Кабинет админа на отдельной странице

AlexanderNaide committed:
Меню в админке

comshepherd commented on BackService/src/main/java/org/satal/backservice/entities/gridClasses/ClubService.java:
```diff
@@ -0,0 +1,35 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id", nullable = false)
+    @Unsigned
```
Это для JDK Flight Recorder?

Есть подоздрение, что это имеет значение только в Event'ах для JFR

AlexanderNaide replied:
Не совсем понял) речь об @Unsigned? Если да, то  это для MySQL, в частности, что использовать диапазон только положительных чисел.

comshepherd replied:
Да, но импорт судя по всему из JDK flight recorder - `import jdk.jfr.Unsigned;`

comshepherd replied:
Можно валидировать Entity перед персистом в базу можно использовать аннотации из `javax.validation` типа `@Positive`

AlexanderNaide replied:
Я пользуюсь MySql, а @Unsigned, кажется встроенная аннотация и с ней работает, без неё нет (ну я так базу настраиваю - не вижу смысла начинать числовые значения id от -2 млрд, если использоваться они будут от 0 и выше).  @Positive у меня отсутствует и я, честно говоря, не разбирался, какую зависимость надо подключить, чтобы этот пакет появился, да и зачем, если @Unsigned делает ровно то, что нужно.

comshepherd commented on BackService/src/main/java/org/satal/backservice/repositories/WorkoutRepository.java:
```diff
@@ -0,0 +1,31 @@
+                    "inner join fitness.users as u\n" +
+                    "on uc.user_id = u.id\n" +
+                    "where yearweek(w.time) = concat(year(curdate()), week(curdate(), 1) + ?1)\n" +
+                    "and u.login = ?2\n" +
```
Необязательно

Лучше использовать  именованные параметры, то есть вместо `?1` использовать `:delta`

Почему?
1. неважно в каком порядке объявлены параметры в методе
2. очень важно, что не нужно переключать взгляд с запроса на параметры в методе, название параметра будет сразу о многом говорить

AlexanderNaide replied:
Хорошая идея, переделаю обязательно.

comshepherd commented on BackService/src/main/java/org/satal/backservice/repositories/WorkoutRepository.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Modifying
+    @Query(nativeQuery = true, value =
+            "select w.id, dayname(w.time) as 'day', s.specialization as 'spec', concat(t.name, ' ', t.surname) as 'trainer', concat(hour(w.time), ':', minute(w.time)) as 'time'\n" +
```
необязательно

1. Чтобы не возвращать String, а более удобную проекцию, можете попробовать вот это - https://thorben-janssen.com/spring-data-jpa-dto-native-queries/
2. Вы возвращаете то что удобно вам, именно такое конкретное представление представление, а можете возвращать сырые данные без преобразования и преобразовывать их уже в ДТО или там где это необходимо, в таком случае получится более универсально

AlexanderNaide replied:
А вот за это спасибо) когда-то я искал такую иформацию. Буду читать, разбираться, надо всё попробовать.
Пока что у меня есть желание наоборот переделать прочие запросы, чтобы не получать сущности, а именно массив строк.
Дело в том, что я могу составить какой угодно запрос в базу чтобы разом получить все необходимые данные для Dto (речь о сложных сущностях, user например, которые в параметрах держат другие сущности). Да, это сложнее, но в работе это 1 запрос всего. А если получать сущности из базы, Page`<User>` например - запрос конечно простой, но при конвертации в Dto автоматически Spring делает по запросу в базу на каждого юзера. А мне этот факт сильно не нравится. По этой же причине я отказался от Specification (сначала делал со спецификациями, на примере магазина, но потом понял, что каждая спека добавляет по отдельному запросу в базу. Придумал универсальный sql запрос, который в принципе делает то же самое: если есть параметр - использует, если нет - пропускает. Это метод findListUsers в UserRepository)

AlexanderNaide committed:
Страница клиентов в админке

AlexanderNaide committed:
Merge pull request #8 from AlexanderNaide/adminAccaunt

Admin accaunt

AlexanderNaide committed:
поиск юзеров без спеки

AlexanderNaide committed:
Merge pull request #9 from AlexanderNaide/adminAccaunt

поиск юзеров без спеки

AlexanderNaide committed:
Исправлен выброс ошибок через Exception. Реализовано перемещение по страницам списка пользователей.

AlexanderNaide committed:
Merge pull request #10 from AlexanderNaide/adminAccaunt

Исправлен выброс ошибок через Exception. Реализовано перемещение по с…

AlexanderNaide committed:
Поиск на фронте

AlexanderNaide committed:
Механизм поиска и фильтрации готов

AlexanderNaide committed:
Merge pull request #11 from AlexanderNaide/adminAccaunt

Admin acaunt

AlexanderNaide committed:
Страница клиентов (условно) закончена.
Пользователям добавлен пол.
В карточку добавлена фотография.
Карточка отображается либо в режиме поиска либо в режиме информации.
Фронт выводит статусы и ошибки в сообщении.

AlexanderNaide committed:
Merge pull request #12 from AlexanderNaide/adminAccaunt

Страница клиентов (условно) закончена.

AlexanderNaide committed:
Новая структура расписания. Переход к повторяющимся позициям

AlexanderNaide committed:
***

AlexanderNaide committed:
Переделал фильтр на несколько параметров (второй параметр пока тестовый)
Добавил кнопку "Добавить к каждому дню".
Исправлена строка времени занятия.

AlexanderNaide committed:
Merge pull request #13 from AlexanderNaide/adminAccaunt

Admin accaunt

AlexanderNaide committed:
Окно получения информации о занятии.
Метод получения из базы конкретного занятия

AlexanderNaide committed:
Merge pull request #14 from AlexanderNaide/adminAccaunt

Окно получения информации о занятии.
